File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/ir/pass/loopforest.cpp
#include "loopforest.h"

namespace IR
{
    void LoopForest::dfsSetLoopDepth(Loop *currentLoop)
    {
        for (auto subLoop : currentLoop->getSubLoops())
        {
            subLoop->setLoopDepth(currentLoop->getLoopDepth() + 1);
            dfsSetLoopDepth(subLoop);
        }
    }

    bool LoopForest::doesAlphaDominatesBeta(BasicBlock *alpha, BasicBlock *beta, const std::map<BasicBlock *, std::set<BasicBlock *>> &dom) const
    {
        auto it = dom.find(beta);
        if (it == dom.end())
        {
            throw std::invalid_argument("BasicBlock not found in DomTree");
        }

        const auto &dominators = it->second;
        return dominators.find(alpha) != dominators.end();
    }

    void LoopForest::dfsFindLoops(Builder &builder, BasicBlock *basicBlock)
    {
        // std::cerr << basicBlock->getIRName() << " DomChildren: " << builder.domTree->getDomChildren(basicBlock).size() << std::endl;
        for (BasicBlock *domSon : builder.domTree->getDomChildren(basicBlock))
        {
            dfsFindLoops(builder, domSon);
        }

        bool isLoopHeader = false;
        std::set<Loop *> subLoops;
        std::set<BasicBlock *> loopBasicBlocks;
        loopBasicBlocks.insert(basicBlock);

        for (BasicBlock *entryBlock : getEntryBlocks(basicBlock))
        {
            if (doesAlphaDominatesBeta(basicBlock, entryBlock, builder.domTree->dom))
            {
                isLoopHeader = true;
                std::queue<BasicBlock *> queue;
                queue.push(entryBlock);

                while (!queue.empty())
                {
                    BasicBlock *u = queue.front();
                    queue.pop();

                    if (basicBlockLoopMap.count(u))
                    {
                        Loop *cloop = basicBlockLoopMap[u];
                        while (cloop->getParentLoop())
                        {
                            cloop = cloop->getParentLoop();
                        }

                        if (!subLoops.count(cloop))
                        {
                            subLoops.insert(cloop);
                            for (auto *block : getEntryBlocks(cloop->getHeaderBasicBlock()))
                            {
                                queue.push(block);
                            }
                        }
                    }
                    else
                    {
                        if (!loopBasicBlocks.count(u))
                        {
                            loopBasicBlocks.insert(u);
                            for (auto *block : getEntryBlocks(u))
                            {
                                queue.push(block);
                            }
                        }
                    }
                }
            }
        }
        // std::cerr << "======================" << std::endl;
        // basicBlock->emitIR(std::cerr);
        if (isLoopHeader)
        {
            Loop *loop = new Loop(basicBlock, subLoops, loopBasicBlocks);
            for (auto subLoop : subLoops)
            {
                subLoop->setParentLoop(loop);
            }
            for (auto *loopBasicBlock : loopBasicBlocks)
            {
                basicBlockLoopMap[loopBasicBlock] = loop;
            }
            basicBlockLoopMap[basicBlock] = loop;
        }
    }

    void LoopForest::initInverseCFG(Function *function)
    {
        auto cfg = function->cfg();
        for (auto &pair : cfg)
        {
            BasicBlock *from = pair.first;
            for (auto &to : pair.second)
            {
                inverse_cfg[to].insert(from);
            }
        }
    }

    std::set<BasicBlock *> LoopForest::getEntryBlocks(BasicBlock *basicBlock)
    {
        if (inverse_cfg.count(basicBlock) == 0)
        {
            return {};
        }
        return inverse_cfg[basicBlock];
    }

    LoopForest::LoopForest(Function *function)
    {
        Builder builder(function);

        initInverseCFG(function);

        dfsFindLoops(builder, function->getEntryBlock());
        // std::cerr << basicBlockLoopMap.size() << std::endl;
        for (const auto &pair : basicBlockLoopMap)
        {
            // auto loop = pair.second;
            // for (auto bb : loop->getBasicBlocks())
            // {
            //     bb->emitIR(std::cerr);
            // }
            if (pair.second->getParentLoop() == nullptr)
            {
                rootLoops.insert(pair.second);
            }
        }
        for (auto rootLoop : rootLoops)
        {
            rootLoop->setLoopDepth(0);
            dfsSetLoopDepth(rootLoop);
        }
    }

    LoopForest::~LoopForest()
    {
        // 清理rootLoops中的Loop对象
        for (auto loop : rootLoops)
        {
            delete loop;
        }
    }
}

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/pass/ir/structure/LoopForest.java
package cn.edu.bit.newnewcc.pass.ir.structure;

import cn.edu.bit.newnewcc.ir.value.BasicBlock;
import cn.edu.bit.newnewcc.ir.value.Function;

import java.util.*;

/**
 * 循环森林
 * <p>
 * 每个函数内包含若干个循环树，这些树构成一个循环森林
 */
public class LoopForest implements Iterable<Loop> {

    /**
     * 构建过程中临时存储的全局信息
     * <p>
     * 将在构建完成后被自动释放
     */
    private static class Builder {
        public final DomTree domTree;

        public Builder(Function function) {
            domTree = DomTree.buildOver(function);
        }
    }

    /**
     * 所有循环树的根
     */
    private final Set<Loop> rootLoops = new HashSet<>();

    /**
     * 基本块到循环的映射关系
     */
    private final Map<BasicBlock, Loop> basicBlockLoopMap = new HashMap<>();

    /**
     * 基于一个函数构建其中基本块形成的循环森林
     *
     * @param function 函数
     */
    private LoopForest(Function function) {
        var builder = new Builder(function);
        dfsFindLoops(builder, function.getEntryBasicBlock());
        for (Loop loop : basicBlockLoopMap.values()) {
            if (loop.getParentLoop() == null) {
                rootLoops.add(loop);
            }
        }
        for (Loop rootLoop : rootLoops) {
            rootLoop.__setLoopDepth__(0);
            dfsSetLoopDepth(rootLoop);
        }
    }

    private void dfsSetLoopDepth(Loop currentLoop) {
        for (Loop subLoop : currentLoop.getSubLoops()) {
            subLoop.__setLoopDepth__(currentLoop.getLoopDepth() + 1);
            dfsSetLoopDepth(subLoop);
        }
    }

    private void dfsFindLoops(Builder builder, BasicBlock basicBlock) {
        for (BasicBlock domSon : builder.domTree.getDomSons(basicBlock)) {
            dfsFindLoops(builder, domSon);
        }
        boolean isLoopHeader = false;
        Set<Loop> subLoops = new HashSet<>();
        Set<BasicBlock> loopBasicBlocks = new HashSet<>();
        loopBasicBlocks.add(basicBlock);
        for (BasicBlock entryBlock : basicBlock.getEntryBlocks()) {
            if (builder.domTree.doesAlphaDominatesBeta(basicBlock, entryBlock)) {
                isLoopHeader = true;
                Queue<BasicBlock> queue = new ArrayDeque<>();
                queue.add(entryBlock);
                while (!queue.isEmpty()) {
                    var u = queue.remove();
                    if (basicBlockLoopMap.containsKey(u)) {
                        var cloop = basicBlockLoopMap.get(u);
                        while (cloop.getParentLoop() != null) {
                            cloop = cloop.getParentLoop();
                        }
                        if (!subLoops.contains(cloop)) {
                            subLoops.add(cloop);
                            queue.addAll(cloop.getHeaderBasicBlock().getEntryBlocks());
                        }
                    } else {
                        if (!loopBasicBlocks.contains(u)) {
                            loopBasicBlocks.add(u);
                            queue.addAll(u.getEntryBlocks());
                        }
                    }
                }
            }
        }
        if (isLoopHeader) {
            var loop = new Loop(basicBlock, subLoops, loopBasicBlocks);
            subLoops.forEach(subLoop -> subLoop.__setParentLoop__(loop));
            loopBasicBlocks.forEach(loopBasicBlock -> basicBlockLoopMap.put(loopBasicBlock, loop));
            basicBlockLoopMap.put(basicBlock, loop);
        }
    }

    /**
     * @return 所有循环树的根循环列表（只读）
     */
    public Collection<Loop> getRootLoops() {
        return Collections.unmodifiableSet(rootLoops);
    }

    /**
     * @return 基本块到循环的映射（只读）
     */
    public Map<BasicBlock, Loop> getBasicBlockLoopMap() {
        return Collections.unmodifiableMap(basicBlockLoopMap);
    }

    @Override
    public Iterator<Loop> iterator() {
        return getRootLoops().iterator();
    }

    /**
     * 在函数的基础上建立一个循环森林 <br>
     * 由于循环标准化的需要，可能对传入的函数产生副作用！！！ <br>
     *
     * @param function 函数（可能会被修改！！！）
     * @return 循环森林
     */
    public static LoopForest buildOver(Function function) {
        return new LoopForest(function);
    }

}
